pkg/comp-functions/functions/spksmariadb/pvcresize.go

```diff
@@ -0,0 +1,260 @@
+		}
+		newSize, found, err := unstructured.NestedString(values, "persistence", "size")
+		if !found {
+			return runtime.NewFatalResult(fmt.Errorf("disk size not found in observed release"))
+		}
+
+		if err != nil {
+			return runtime.NewFatalResult(fmt.Errorf("failed to read size from observed release: %w", err))
+		}
+		desiredSize, err := getSizeAsInt(newSize)
+		if err != nil {
+			return runtime.NewFatalResult(fmt.Errorf("cannot parse desired size: %w", err))
+		}
+		stsUpdated := stsSize == desiredSize
+
+		deletionJob := &batchv1.Job{}
+		if observedJob {
+			err := json.Unmarshal(xJob.Status.AtProvider.Manifest.Raw, deletionJob)
+			if err != nil {
+				return runtime.NewFatalResult(fmt.Errorf("cannot unmarshal sts deleter job: %w", err))
+			}
+		}
+
+		// The job hasn't been observed yet, so we need to keep it in desired, or we will have a recreate loop
+		// Also as long as it hasn't finished we need to make sure it exists.
+		if (!observedJob || deletionJob.Status.Succeeded < 1) || (sts.Status.ReadyReplicas == 0 && !stsUpdated) {
+			err := addDeletionJob(svc, comp, newSize, release.GetName())
+			if err != nil {
+				return runtime.NewFatalResult(fmt.Errorf("cannot create RBAC for the deletion job: %w", err))
+			}
+		}
+		return nil
+	}
+
+	err = addStsObserver(svc, comp)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Errorf("cannot observe sts: %w", err).Error())
+	}
+
+	sts := &appsv1.StatefulSet{}
+	err = svc.GetObservedKubeObject(sts, comp.Name+"-sts-observer")
+	if err == runtime.ErrNotFound {
+		return nil
+	}
+	if err != nil {
+		return runtime.NewFatalResult(fmt.Errorf("cannot get observed statefulset: %w", err))
+	}
+	stsSize := int64(0)
+	// Check the current size in the sts
+	if len(sts.Spec.VolumeClaimTemplates) > 0 {
+		stsSize, _ = sts.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage().AsInt64()
+	}
+
+	patch, newSize, result := needReleasePatch(values, stsSize)
+	if result != nil {
+		return result
+	}
+
+	if patch {
+		err = addDeletionJob(svc, comp, newSize, release.GetName())
+		if err != nil {
+			return runtime.NewFatalResult(fmt.Errorf("cannot create the deletion job: %w", err))
+		}
+		// We pause the release at this point to make sure that provider-helm doesn't update the
+		// release until the deletion job removed the sts
+		release.SetAnnotations(map[string]string{
+			"crossplane.io/paused": "true",
+		})
+
+		err = svc.SetDesiredComposedResourceWithName(release, mariadbRelease)
+		if err != nil {
+			return runtime.NewFatalResult(fmt.Errorf("Can't pause the release: %w", err))
+		}
+		return nil
+	}
+
+	return nil
+}
+
+func addStsObserver(svc *runtime.ServiceRuntime, comp *spksv1alpha1.CompositeMariaDBInstance) error {
+	statefulset := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "mariadb",
+			Namespace: comp.GetName(),
+		},
+	}
+
+	providerConfigRef := comp.GetLabels()["service.syn.tools/cluster"]
```

Comment:
I'd add here a check if this key exists in map otherwise You'll be working with an empty string